docs/source/app-dev/command-deduplication.rst

```diff
@@ -0,0 +1,339 @@
+
+
+.. _dedup-bounded-error-handling:
+
+Error handling
+--------------
+
+Error handling is needed when the status code of the RPC call or in the :ref:`in the completion event <com.daml.ledger.api.v1Completion.status>` is not ``OK``.
+The following table lists appropriate reactions by status code (written as ``STATUS_CODE``) and error code (written in capital letters with a link to the error code documentation).
+Fields in the error metadata are written as ``field`` in lowercase letters.
+
+.. list-table:: Command deduplication error handling with known processing time bound
+   :widths: 10 50
+   :header-rows: 1
+
+   - * Error condition
+     
+     * Reaction
+
+       
+   - * ``DEADLINE_EXCEEDED``
+     
+     * Consider the submission as lost.
+       
+       Retry from obtaining a completion offset ``OFF1`` (:ref:`step 2 <dedup-bounded-step-offset>`) and possibly increase the timeout.
+
+       
+   - * Application crashed
+     
+     * Retry from obtaining a completion offset (:ref:`step 2 <dedup-bounded-step-offset>`).
+
+
+   - * ``ALREADY_EXISTS`` / :ref:`DUPLICATE_COMMAND <error_code_DUPLICATE_COMMAND>`
+     
+     * The change ID has already been accepted on the ledger within the reported deduplication period.
+       The optional field ``completion_offset`` contains the precise offset.
+       Report success for the ledger change.
+       
+       If desired, query the ``completion_offset`` via the :ref:`Command Completion Service <command-submission-service>` to find out about the earlier outcome.
+
+       
+   - * ``FAILED_PRECONDITION`` / :ref:`INVALID_DEDUPLICATION_PERIOD <error_code_INVALID_DEDUPLICATION_PERIOD>`
+     
+     * The specified deduplication duration is longer than what the Daml ledger supports.
+       ``earliest_offset`` contains the earliest deduplication offset or ``longest_duration`` contains the longest deduplication duration that can be used (at least one of the two must be provided).
+
+       Options:
+
+       - Negotiate support for longer deduplication periods with the ledger operator.
+
+       - Set the deduplication offset to ``earliest_offset`` or the deduplication duration to ``longest_duration`` and retry from obtaining a completion offset (:ref:`step 2 <dedup-bounded-step-offset>`).
+	 This may lead to accepting the change twice.
+
+	 
+   - * ``ABORTED`` / :ref:`SUBMISSION_ALREADY_IN_FLIGHT <error_code_SUBMISSION_ALREADY_IN_FLIGHT>`
+     
+       This error occurs only as an RPC response, not inside a completion event.
+       
+     * There is already another submission in flight, with the submission ID in ``existing_submission_id``.
+
+       - When you use the :ref:`Command Service <command-service>`, wait a bit and retry from submitting the command (:ref:`step 3 <dedup-bounded-step-submit>`).
```

Comment:
It is not clear that the submission has already been accepted; it just says that there is an in-flight submission, but this submission could still be rejected. This can also happen with the command service. For example:
1. Submit via the command service with a very short gRPC deadline. The command service responds with `DEADLINE_EXCEEDED`.
2. You immediately resubmit with a higher deadline. This gives you `SUBMISSION_ALREADY_IN_FLIGHT`, as the submission from step 1 is still being processed. (This probably doesn't happen with kv based ledgers where the command deduplication check is atomic with the commit, but it can happen in Canton where the two are not atomic.)
3. Since the gRPC call from step 1 has exceeded the deadline, you don't know the outcome of the submission. So you wait a bit until you have some confidence that the earlier submission should now be processed. Alas, you don't want to use the completion service. So instead you submit the command again. If the submission from step 1 was accepted, then you'll get back ALREADY_EXISTS. If it failed, you now have a second chance to get it accepted by the ledger and learn the outcome.